Multicluster concept doc (#2991)
istio/istio.io: content/docs/concepts/multicluster-configuration/index.md

---

> title: Multicluster Istio

Reviewer: How about "Multicluster Deployments" instead? "Istio" is kind redundant.

essentially combining the microservices into a controllable composite application.
For applications up to a certain size, all of the microservices comprising the
application can be running on a single orchestration platform (e.g., kubernetes cluster),
but for many reasons (e.g., scale, redundancy, etc.) most applications will eventually
need to be distributed and have some of their services running elsewhere.

Istio supports many possible topologies for distributing the services of an
application beyond a single cluster.
Services within the mesh can use [service entries](/docs/concepts/traffic-management/#service-entries)
to access standalone external services or services exposed by another loosely-coupled
service mesh, a.k.a. *mesh federation*.
Alternatively, if all the services can be under the same administrative control,
you can [expand the service mesh](/docs/setup/kubernetes/mesh-expansion/) to include services running
on VMs or bare metal hosts, or you can combine the services from more than one cluster into a
single composite service mesh, i.e., a *multicluster mesh*.

## Multicluster service mesh

> A multicluster service mesh is defined as a single composite service mesh composed of services

Reviewer: composite & composed, doesn't sound too good. Maybe replace "composed" with "made up of"?

under a single administrative control.
A service named `foo` in cluster 1 is the same service as `foo` in cluster 2. This is
unlike a loosely-coupled federated service mesh where two clusters may have different
definitions of the same service, which need to be reconciled when integrating them.

A multicluster mesh has the advantage that all the services look the same to clients,
regardless of where the workloads are actually running. It's transparent
to the application whether it's deployed in a single or multicluster mesh.
To achieve this, all of the services need to be managed by a single logical
control plane, although not necessarily a single physical Istio control plane.
There are two possible deployment approaches:

1. A single Istio control plane with access to and configuration for all the services in the mesh.

1. Multiple synchronized Istio control planes with replicated service and routing configurations. 

Even within these two topologies, there is more than one way to configure things.
Which approach to use and how to configure it depends on the requirements of the application
as well as the features and limitations of the underlying cloud deployment platform.

> ### Single control plane topology

Reviewer 2: You need to add the biggest caveat here that its the users responsibility to setup DNS resolution for services in remote clusters (as we dont have service entries anymore). While the multicluster one comes with DNS resolution for services

This multicluster configuration uses a single Istio control plane,
typically running on one of the clusters, whose Pilot
manages services on both local and remote clusters and configures the Envoy sidecars
for all of them. This approach works best in environments where all of the contributing
clusters have VPN connectivity so every pod in the mesh is reachable from anywhere else using
the same IP address.

{{< image width="80%" ratio="36.01%"

>     link="./multicluster-with-vpn.svg"

Reviewer 3: Thank you for including diagrams. However, this diagram is overloaded with text making it very difficult to understand. Consider simplifying the diagram and removing the numbered steps from the diagram and instead including them in an ordered list in the text. You can then only have the numbers appear in the diagram making it easier to understand. Evaluate if you need to include all the technical details in the diagram. Ask yourself whether or not the readers need those technical detail labels to understand the **concept** you are trying to convey. Lastly, the lack of a legend leaves readers wondering what meaning the different types of shapes and lines have. Consider adding a legend to make their meaning clear.

    caption="Istio mesh spanning multiple Kubernetes clusters with direct network access to remote pods over VPN"
    >}}

In this configuration, one or more clusters run a remote Istio configuration which connects them
to a single Istio control plane, allowing it to manage all of the Envoy's as a single mesh.
Detailed instructions for setting up this kind of multicluster topology can be found
in [single control plane with vpn](/docs/setup/kubernetes/multicluster-install/vpn/).

If setting up a VPN network to provide universal pod-to-pod connectivity is difficult or impossible,
Istio Pilot can be configured to enable location-aware service routing (a.k.a. split-horizon EDS).
When configured this way, a request from a sidecar in one cluster to a service that is
also running in the same cluster will be forwarded to the local service IP as usual, but when the destination
workload is running in a different cluster it will use the remote cluster Gateway IP to connect to the service instead.

> {{< image width="80%" ratio="36.01%"

Reviewer 3: Consider writing a brief introduction to the diagram for accessibility purposes.

>     link="./multicluster-split-horizon-eds.svg"

Reviewer 3: Similarly to the previous diagram, this one is also overloaded with technical details which are not really necessary to convey the concept you are trying to communicate and it is also missing a legend to know the meaning of the different types and colors of shapes and lines. Consider simplifying the diagram by providing a higher level view of the topology and including only the technical details that are absolutely necessary. Having a legend instead of labeling every line, for example, goes a long way in making diagrams easier to understand for readers.

    caption="Istio mesh spanning multiple Kubernetes clusters using single Istio control plane and Gateway to reach remote pods"
    >}}

The [single control plane with gateways](/docs/examples/multicluster/split-horizon-eds/) example provides

> instructions for experimenting with this feature.

Reviewer 3: Consider changing this sentence to a call-to-action to motivate readers to follow the link:
```suggestion
Visit our [instructions on using the single control plane with gateways](/docs/examples/multicluster/split-horizon-eds/) to experiment with this feature.
```

### Multiple control plane topology

Instead of using a central Istio control plane to manage the mesh,
in this configuration each cluster has an **identical** Istio control plane

> installation, each managing its own endpoints.

Reviewer 3: At this point, the text has already established that the two topologies are different and how they differ. The conditional clause at the beginning of the sentence makes it about the contrast instead of about defining the multiple control plane topology's characteristics. Consider:
```suggestion
In a multiple control plane configuration, each cluster has an **identical** Istio control plane
installation and each control plane manages its own endpoints.
```

> All of the clusters are under a shared administrative control for the purposes of

Reviewer 4: I'd rework this sentence as follows
`All of the clusters share administrative control of Istio policy enforcement and security configuration.`

> policy enforcement and security.

Reviewer 3: Consider simplifying the text by removing unnecessary words to make it easier to understand for readers with ESL.
```suggestion
All the clusters are under a shared administrative control for policy enforcement and security.
```

> {{< image width="80%" ratio="36.01%"

Reviewer 3: Consider writing a brief introduction to the diagram for accessibility.

>     link="./multicluster-with-gateways.svg"

Reviewer 3: Please see my previous comments regarding the diagram on line 60. The same observations apply.

    caption="Istio mesh spanning multiple Kubernetes clusters using multiple Istio control planes and Gateway to reach remote pods"
    >}}

> A single Istio service mesh across the clusters is achieved by replicating

Reviewer 4: this is a good place to put in "A single logical service mesh`

> shared services and namespaces and using a common root CA in all of the clusters.

Reviewer 3: The sentence raises the question in the reader's mind: Who or what can achieve the single mesh?
It also leaves CA undefined. Consider instead:
```suggestion
To achieve a single Istio service mesh across the clusters, you must use a common root Certificate Authority (CA) and replicate the shared services and namespaces in all clusters.
```

Reviewer 4: This is true of either mc technology and isn't a distinguishing fact.

> Cross-cluster communication occurs over Istio Gateways of the respective clusters.

Reviewer 3: Please add an article to make the text read better:
```suggestion
Cross-cluster communication occurs over the Istio gateways of the respective clusters.
```

> 

Author: TODO: mention split-horizon applicability here too?

Check out [multiple control plane with gateways](/docs/setup/kubernetes/multicluster-install/gateways/)

> for instructions on setting up this this kind of multicluster configuration.

Reviewer 3: This text has a clear call-to-action, which is good, but it gives readers the contents at the end. Thus, readers could click the link without knowing whether or not it is really something they want to do. Consider reversing the clauses:
```suggestion
To set up this this kind of multicluster configuration, visit the [installation instructions for multiple control planes with gateways](/docs/setup/kubernetes/multicluster-install/gateways/).
```
